evidence/verify.go

```diff
@@ -1,79 +1,257 @@
-func VerifyEvidence(evidence types.Evidence, state sm.State, stateDB StateStore, blockStore BlockStore) error {
+// - it is internally consistent with state
+// - it was properly signed by the alleged equivocator and meets the individual evidence verification requirements
+func (evpool *Pool) verify(evidence types.Evidence) (*Info, error) {
 	var (
+		state          = evpool.State()
 		height         = state.LastBlockHeight
 		evidenceParams = state.ConsensusParams.Evidence
-
-		ageDuration  = state.LastBlockTime.Sub(evidence.Time())
-		ageNumBlocks = height - evidence.Height()
-
-		header *types.Header
+		ageNumBlocks   = height - evidence.Height()
 	)
 
-	// if the evidence is from the current height - this means the evidence is fresh from the consensus
-	// and we won't have it in the block store. We thus check that the time isn't before the previous block
-	if evidence.Height() == height+1 {
-		if evidence.Time().Before(state.LastBlockTime) {
-			return fmt.Errorf("evidence is from an earlier time than the previous block: %v < %v",
-				evidence.Time(),
-				state.LastBlockTime)
-		}
-	} else {
-		// try to retrieve header from blockstore
-		blockMeta := blockStore.LoadBlockMeta(evidence.Height())
-		header = &blockMeta.Header
-		if header == nil {
-			return fmt.Errorf("don't have header at height #%d", evidence.Height())
-		}
-		if header.Time != evidence.Time() {
-			return fmt.Errorf("evidence time (%v) is different to the time of the header we have for the same height (%v)",
-				evidence.Time(),
-				header.Time,
-			)
-		}
+	// check that the evidence isn't already committed
+	if evpool.isCommitted(evidence) {
+		return nil, errors.New("evidence was already committed")
+	}
+
+	// verify the time of the evidence
+	blockMeta := evpool.blockStore.LoadBlockMeta(evidence.Height())
+	if blockMeta == nil {
+		return nil, fmt.Errorf("don't have header at height #%d", evidence.Height())
 	}
+	evTime := blockMeta.Header.Time
+	ageDuration := state.LastBlockTime.Sub(evTime)
 
+	// check that the evidence hasn't expired
 	if ageDuration > evidenceParams.MaxAgeDuration && ageNumBlocks > evidenceParams.MaxAgeNumBlocks {
-		return fmt.Errorf(
+		return nil, fmt.Errorf(
 			"evidence from height %d (created at: %v) is too old; min height is %d and evidence can not be older than %v",
 			evidence.Height(),
-			evidence.Time(),
+			evTime,
 			height-evidenceParams.MaxAgeNumBlocks,
 			state.LastBlockTime.Add(evidenceParams.MaxAgeDuration),
 		)
 	}
 
-	valset, err := stateDB.LoadValidators(evidence.Height())
+	// apply the evidence-specific verification logic
+	switch ev := evidence.(type) {
+	case *types.DuplicateVoteEvidence:
+		valSet, err := evpool.stateDB.LoadValidators(evidence.Height())
+		if err != nil {
+			return nil, err
+		}
+		err = VerifyDuplicateVote(ev, state.ChainID, valSet)
+		if err != nil {
+			return nil, fmt.Errorf("verifying duplicate vote evidence: %w", err)
+		}
+
+		_, val := valSet.GetByAddress(ev.VoteA.ValidatorAddress)
```

Comment:
should we check that val is not nil here?

Reply:
We already checked that it was in the validator set in `VerifyDuplicateVote`